### PR TITLE
Preserve gradient checkpointing kwargs during GKD generation

### PR DESCRIPTION
**Pull request by DimensionSTP**
# What does this PR do?

GKD temporarily disables gradient checkpointing while unwrapping the student or teacher model for generation. The previous restoration path called `gradient_checkpointing_enable()` without the configured `gradient_checkpointing_kwargs`.

With reentrant checkpointing under DeepSpeed ZeRO-3, this changes checkpointing behavior between the original forward and backward recomputation. The recomputation can then observe partitioned zero-sized parameters and fail with:

```text
torch.utils.checkpoint.CheckpointError:
Recomputed values have different metadata than during the forward pass.
saved shape: [8192, 2560]
recomputed shape: [0]
```

This change:

- lets the generation unwrap context receive the original checkpointing kwargs;
- restores checkpointing through the existing `disable_gradient_checkpointing` context manager, including exception exits;
- passes `self.args.gradient_checkpointing_kwargs` from both GKD generation paths;
- adds focused tests for normal and exceptional restoration.

The GKD objective, generation behavior, and DeepSpeed gathering policy are unchanged.

## Verification

- Two-rank DeepSpeed ZeRO-3 reproducer before the fix: deterministic checkpoint metadata mismatch
- Same reproducer after the fix: one optimizer step completed
- `pytest -q tests/test_model_utils.py`: 7 passed
- Focused restoration tests: 2 passed
- Ruff check for the changed tests: passed
- Ruff format check for all three changed files: passed
- `compileall`: passed
- `git diff --check`: passed

## Before submitting

- [x] Did you read the contributor guidelines?
- [ ] Was this discussed/approved via a GitHub issue? Not applicable; this is a self-contained bug fix with a minimal reproducer.
- [x] Did you make sure to update the documentation? No user-facing documentation change is required.
- [x] Did you write the necessary tests?

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

---

> [!NOTE]
> **Medium Risk**
> Touches the training generation unwrap path and gradient-checkpointing restore logic used with DeepSpeed ZeRO-3; behavior is narrowed to matching pre-generation checkpointing settings.
> 
> **Overview**
> Fixes a bug where **GKD** temporarily disables gradient checkpointing for on-policy generation but re-enabled it **without** the original `gradient_checkpointing_kwargs` (e.g. `use_reentrant`), which can break **DeepSpeed ZeRO-3** training with `CheckpointError` metadata mismatches.
> 
> **`unwrap_model_for_generation` / `_unwrap_model_for_generation`** now accept optional `gradient_checkpointing_kwargs` and delegate disable/restore to **`disable_gradient_checkpointing`**, so checkpointing settings are restored on normal exit and after exceptions. **`disable_gradient_checkpointing`** passes kwargs explicitly as `gradient_checkpointing_enable(gradient_checkpointing_kwargs=...)`.
> 
> **`GKDTrainer.training_step`** forwards `self.args.gradient_checkpointing_kwargs` for both student and teacher generation unwrap paths. New unit tests cover kwargs restoration and post-exception cleanup.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f2ef118b9e3ee5bcc165b65c5ccfa113b74b047e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/models/utils.py [hunk 1]
```diff
@@ -131,25 +134,21 @@ def _unwrap_model_for_generation(
...
-        unwrapped_model.gradient_checkpointing_enable()
+            else:
+                import deepspeed
+
+                with deepspeed.zero.GatheredParameters(model.parameters()):
+                    remove_hooks(model)
+                    yield accelerator.unwrap_model(model)
+                    add_hooks(model)
+        else:
+            yield unwrapped_model
```
[the remark this answers is not in the document]
Comment (DimensionSTP) [reply trl/models/utils.py:151]: Confirmed, thank you. Fixed in 38f9dcc by restoring `gradient_checkpointing_kwargs` as a keyword argument. The focused tests now use a kwargs-only wrapper contract for both normal and exceptional exits, and an actual `PolicyAndValueWrapper` smoke passes.

**Comment by DimensionSTP** [comment]
@qgallouedec, could you take a look when you have a chance?

I rebased this PR onto current main (`94d6b7dd`), keeping the fix that restores gradient-checkpointing kwargs by keyword. The scope remains the generation unwrap context and the GKD call sites.

Revalidation:
- `tests/test_model_utils.py`: 7 passed, including restoration after an exception.
- The existing 2-rank ZeRO-3 GKD reproducer completed a training step with reentrant gradient checkpointing.
- Ruff lint/format and `git diff --check`: passed.

The distributed run used Transformers 5.15.1, Torch 2.11.0 and DeepSpeed 0.18.8. This is a focused GKD regression check, not a claim of full DistillationTrainer coverage. Thanks!

**Comment by DimensionSTP** [comment]
Merged the latest main (`e47e6d98`) into this PR. The conflict was limited to a docstring; both the upstream formatting update and the checkpointing-kwargs documentation are preserved.

Validation: all 7 model-utils CPU tests pass, along with Ruff lint/format, compilation, and diff checks. GPU tests were not rerun for this merge.

Updated head: `f2ef118b`. GitHub now reports no merge conflicts.

**Comment by DimensionSTP** [comment]
Hi @qgallouedec - I refreshed this branch against main on September 10, and it is still mergeable with green checks.

The patch only preserves the original gradient-checkpointing kwargs while GKD temporarily unwraps the model for generation. It does not alter checkpointing policy or the generation flow itself.

Could you take a look when you have a slot? In particular, I'd like to confirm that restoring the kwargs by keyword is consistent with the model-utils contract.

**Comment by DimensionSTP** [comment]
Coordination with #7162: @YeonwooSung [confirmed](https://github.com/huggingface/trl/pull/7162#issuecomment-5787541091) that #7162 handles exception-safe checkpointing restoration and ZeRO-3 hook cleanup, while this PR remains focused on preserving configured `gradient_checkpointing_kwargs` through the unwrap context and the GKD student/teacher generation call sites. These fixes are complementary, not duplicates.

I previously checked the combined behavior in a 72-case CPU mock matrix; that was not a new distributed GPU run. If #7162 lands first, I will update this PR on top of it, retain its inner `add_hooks` `finally`, and rerun the focused model-utils and two-rank ZeRO-3 checks before requesting merge.
